### PR TITLE
Add Ruby 3 breaking changes section to upgrade guides

### LINKED ISSUE #41270: wrong number of arguments (given 3, expected 1..2) on errors.add when given hash
### Steps to reproduce

Use errors.add in your model to add validation errors:

```rb
class User < ApplicationRecord
  def add_error
    errors.add(:name, :invalid, {})
  end
end

User.new.add_error
```

### Expected behavior

Not throwing an error (which is Rails 6.0 behaviour)

### Actual behavior

`wrong number of arguments (given 3, expected 1..2)` on Rails 6.1

This boils down to a changed method signature between 6 and 6.1

6.0: https://api.rubyonrails.org/v6.0.0/classes/ActiveModel/Errors.html#method-i-add
6.1: https://api.rubyonrails.org/v6.1.1/classes/ActiveModel/Errors.html#method-i-add

The change was introduced in a three years old commit: https://github.com/rails/rails/commit/d9011e39357300fe78720227af4c13b4bc4ac4dd (quite some archeology here 🦖 )

The fix for consumer side is rather easy:

```rb
errros.add(:name, invalid, **{})
```

Not sure what the Rails team can do here, but that breaking change definitely caught me by surprise on one project, since it's not mentioned at all in the [6.1 changelog for ActiveModel](https://github.com/rails/rails/blob/6-1-stable/activemodel/CHANGELOG.md). If nothing else, this issue post here should allow people to find if they run into it (which they will with current versions of kt-paperclip, since this is how I found the issue)

Thanks for reading and keep on the good work!

### System configuration
**Rails version**: 6.1.1

**Ruby version**: 3.0.0

### PR DESCRIPTION
**Pull request by intrip**
### Summary

Add a new section to upgrade guides which outlines the Ruby 3 breaking changes and how such changes affects Rails API usage.

Closes #41270

### CODE REVIEW

**Comment by intrip** [comment]
@zzak can you take a peek 🙏 ? 
I hope the ping-pong won't take too much 😆

File: guides/source/upgrading_ruby_on_rails.md [hunk 1]
```diff
@@ -26,6 +26,38 @@ Rails generally stays close to the latest released Ruby version when it's releas
...
+User.new.add_error
+# Works in Ruby <= 2.7, raises ArgumentError in Ruby 3
+```
+
+After:
+
+```ruby
+class User < ApplicationRecord
+  def add_error
+    errors.add(:name, :invalid, **{ message: "ops there was an error." })
```
Comment (sunny) [inline guides/source/upgrading_ruby_on_rails.md:53]: Should’nt we encourage the simplified syntax without the splats instead, here?

Maybe something like:

```suggestion
    # Equivalent to:
    #   errors.add(:name, :invalid, **{ message: "ops there was an error."})
    errors.add(:name, :invalid, message: "ops there was an error.")
```

File: guides/source/upgrading_ruby_on_rails.md [hunk 2]
```diff
@@ -26,6 +26,38 @@ Rails generally stays close to the latest released Ruby version when it's releas
...
+Ruby 3 introduces breaking changes to [keyword arguments](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/). Whilst we do our best in order to make sure the Rails API is compatible between supported Ruby versions sometimes this isn't possible, below you'll find some (but not all) examples of such issues.
+
+Passing the keyword argument as the last hash parameter is no longer working, you need to transform the hash parameter to keywords by using the double splat operator `**`:
+
+Before:
+
+```ruby
+class User < ApplicationRecord
+  def add_error
+    errors.add(:name, :invalid, { message: "ops there was an error." })
```
Comment (sunny) [inline guides/source/upgrading_ruby_on_rails.md:40]: I’d suggest using an example message that works well with the Rails way: that supposes it will be prefixed by the column name and not ending with a period like other messages so as to work with `to_sentence`. E.g.:

```suggestion
    errors.add(:name, :invalid, { message: "is not allowed" })
```

File: guides/source/upgrading_ruby_on_rails.md [hunk 3]
```diff
@@ -26,6 +26,38 @@ Rails generally stays close to the latest released Ruby version when it's releas
 
 It's a good idea to upgrade Ruby and Rails separately. Upgrade to the latest Ruby you can first, and then upgrade Rails.
 
+#### Ruby 3 breaking changes
+
+Ruby 3 introduces breaking changes to [keyword arguments](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/). Whilst we do our best in order to make sure the Rails API is compatible between supported Ruby versions sometimes this isn't possible, below you'll find some (but not all) examples of such issues.
+
+Passing the keyword argument as the last hash parameter is no longer working, you need to transform the hash parameter to keywords by using the double splat operator `**`:
```
Comment (sunny) [inline guides/source/upgrading_ruby_on_rails.md:33]: Some minor suggestions to simplify the wording slightly. Please take this with a grain of salt as this is mostly a matter of personal preference. 🤗 

```suggestion
Ruby 3 introduces breaking changes to [keyword arguments](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/). While we do our best in order to make sure the Rails API is compatible between supported Ruby versions sometimes this isn't possible. Below you'll find an example of such an issue.

Passing the keyword argument as the last hash parameter is no longer allowed. You need to use keywords or transform the hash parameter to keywords by using the double splat operator `**`:
```

File: guides/source/upgrading_ruby_on_rails.md (hunk 1 above)
Comment (ghiculescu) [reply guides/source/upgrading_ruby_on_rails.md:53]: Yes please ^ that syntax works across versions doesn’t it?

**Comment by eileencodes** [comment]
Personally I don't think this changelog/documentation about Ruby 3.0 should be included. Ruby 2.7 provided all the warnings necessary to fix these deprecations before upgrading to 3.0. We haven't in the past provided docs about other Ruby changes that someone might have not fixed when upgrading. There is so much application code that could break during the 2.7 to 3.0 upgrade and since we're not going to cover all the possible breakage I don't think we should include docs about the active model errors.

File: guides/source/upgrading_ruby_on_rails.md (hunk 1 above)
Comment (sunny) [reply guides/source/upgrading_ruby_on_rails.md:53]: Oh yeah, I think that’s even clearer!

**Comment by intrip** [comment]
Thanks for the quick feedback @eileencodes!

Any guidance on how we could proceed about https://github.com/rails/rails/issues/41270 ?
Basically between 6.0 and 6.1 there was an API changes which introduced a Ruby3 issue: from the above comment I understand that we should add a changelog line ?

**Comment by eileencodes** [comment]
The issue links to the commit that made changes to ActiveModel::Error which was documented in the CHANGELOG here https://github.com/rails/rails/pull/36125. I think that covers the concerns in #41270 - maybe we can make the existing changelog clearer.

File: guides/source/upgrading_ruby_on_rails.md (hunk 3 above)
Comment (p8) [reply guides/source/upgrading_ruby_on_rails.md:33]: I think the "in order" can be removed...
"While we do our best to make sure the Rails API is compatible..."

**Comment by intrip** [comment]
@eileencodes created https://github.com/rails/rails/pull/42640 as follow up 🙇
